### ISSUE TITLE
How can I deserialize this untagged enum?

### ISSUE THREAD

**Issue by phlopsi** [serde-rs/serde#1678]
I'm currently writing my own deserializer for an external custom format. It looks something like this:
```
ABC
BCD=CDE
DEF={
  EFG
  FGH="some random text"
  {
    "something different"=GHI
  }
}
```
I decided to go with the following structure for deserialization:
```rust
// parser.rs
use serde::Deserialize;

#[derive(Deserialize)]
pub struct ExpressionStream<'a>(#[serde(borrow)] Vec<Expression<'a>>);

#[derive(Deserialize)]
#[serde(untagged)]
pub enum Expression<'a> {
    KeyValuePair(KeyValuePair<'a>),
    Group(ExpressionStream<'a>),
    Literal(&'a [u8]),
}

#[derive(Deserialize)]
pub struct KeyValuePair<'a>(&'a [u8], Value<'a>);

#[derive(Deserialize)]
#[serde(untagged)]
pub enum Value<'a> {
    Group(ExpressionStream<'a>),
    Literal(&'a [u8]),
}
```
In order to deserialize my input, I call `ExpressionStream::deserialize(&mut Deserializer::new(&input[..]))`. At the moment, I have problems with deserializing the `KeyValuePair` enum.

Here's my `deserialize_any`-method:
```rust
    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        let peek = match self.parse_whitespace()? {
            Some(b) => b,
            None => {
                return Err(Error);
            }
        };

        if b'{' == peek {
            unimplemented!()
        } else if b'"' == peek {
            unimplemented!()
        } else if IDENTIFIER_CHARACTER_CATEGORY[peek as usize] {
            let identifier = self.parse_identifier()?;

            match self.parse_whitespace()? {
                Some(b'=') => {
                    // Remove the equals sign from the input
                    self.next();

                    match self.parse_whitespace()? {
                        Some(b'{') => unimplemented!(),
                        Some(b'"') => unimplemented!(),
                        // This is called for key-value pairs where both the left-hand
                        // and right-hand side are literals without quotation marks.
                        Some(_peek) => unimplemented!(),
                        None => Err(Error),
                    }
                }
                // This is called for literals without quotation marks
                _ => visitor.visit_borrowed_bytes(identifier),
            }
        } else {
            unimplemented!()
        }
    }
```
The branch I'm stuck at is this one:
```rust
                        // This is called for key-value pairs where both the left-hand
                        // and right-hand side are literals without quotation marks.
                        Some(_peek) => unimplemented!(),
```
From what I understand, I'd have to call `visitor.visit_seq([...])` to create a `Content::Seq` which is needed for the deserializer to create a `KeyValuePair`. I'm thinking about creating a new struct implementing the `SeqAccess` trait, that already holds the key and value and simply returns them in order when calling `next_element`, but that's where I'm lost. I have neither an idea how to achieve that nor do I know if that's even the correct approach.

**Comment by phlopsi** [comment]
I experimented a bit and tried the following:
```rust
                        Some(_peek) => visitor.visit_seq(SeqDeserializer::new(
                            vec![
                                visitor.visit_borrowed_bytes(identifier)?,
                                visitor.visit_borrowed_bytes(self.parse_identifier()?)?,
                            ]
                            .into_iter(),
                        )),
```
However, that results in the following error:
```
error[E0277]: the trait bound `<V as parser::_IMPL_DESERIALIZE_FOR_ExpressionStream::_serde::de::Visitor<'de>>::Value: parser::_IMPL_DESERIALIZE_FOR_ExpressionStream::_serde::
de::IntoDeserializer<'_, _>` is not satisfied
   --> src\de.rs:283:58
    |
283 |                           Some(_peek) => visitor.visit_seq(SeqDeserializer::new(
    |  __________________________________________________________^
284 | |                             vec![
285 | |                                 visitor.visit_borrowed_bytes(identifier)?,
286 | |                                 visitor.visit_borrowed_bytes(self.parse_identifier()?)?,
287 | |                             ]
288 | |                             .into_iter(),
289 | |                         )),
    | |_________________________^ the trait `parser::_IMPL_DESERIALIZE_FOR_ExpressionStream::_serde::de::IntoDeserializer<'_, _>` is not implemented for `<V as parser::_IMPL_D
ESERIALIZE_FOR_ExpressionStream::_serde::de::Visitor<'de>>::Value`
    |
    = note: required because of the requirements on the impl of `parser::_IMPL_DESERIALIZE_FOR_ExpressionStream::_serde::de::SeqAccess<'de>` for `parser::_IMPL_DESERIALIZE_FOR
_ExpressionStream::_serde::de::value::SeqDeserializer<std::vec::IntoIter<<V as parser::_IMPL_DESERIALIZE_FOR_ExpressionStream::_serde::de::Visitor<'de>>::Value>, _>`
```

**Comment by phlopsi** [comment]
I added my own `KeyValuePairDeserializer` struct, that implements both `Deserializer` and `SeqAccess`:
```rust
enum KeyValuePairDeserializationState {
    DeserializedNothing,
    DeserializedKey,
    DeserializedValue,
}

struct KeyValuePairDeserializer<'a, 'de: 'a> {
    deserializer: &'a mut Deserializer<'de>,
    key: &'de [u8],
    state: KeyValuePairDeserializationState,
}

impl<'a, 'de> KeyValuePairDeserializer<'a, 'de> {
    pub fn new(deserializer: &'a mut Deserializer<'de>, key: &'de [u8]) -> Self {
        Self {
            deserializer,
            key,
            state: KeyValuePairDeserializationState::DeserializedNothing,
        }
    }
}

impl<'de, 'a, 'b> de::Deserializer<'de> for &'a mut KeyValuePairDeserializer<'b, 'de> {
    type Error = Error;

    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_borrowed_bytes(self.key)
    }

    // [...]
}

impl<'de, 'a> SeqAccess<'de> for KeyValuePairDeserializer<'a, 'de> {
    type Error = Error;

    fn next_element_seed<T>(&mut self, seed: T) -> Result<Option<T::Value>>
    where
        T: DeserializeSeed<'de>,
    {
        match self.state {
            KeyValuePairDeserializationState::DeserializedNothing => {
                self.state = KeyValuePairDeserializationState::DeserializedKey;
                seed.deserialize(&mut *self).map(Some)
            }
            KeyValuePairDeserializationState::DeserializedKey => {
                self.state = KeyValuePairDeserializationState::DeserializedValue;
                seed.deserialize(&mut *self.deserializer).map(Some)
            }
            KeyValuePairDeserializationState::DeserializedValue => Ok(None)
        }
    }
}
```
I've also edited my regular deserializer:
```rust
pub struct Deserializer<'de> {
    input: &'de [u8],
    is_deserializing_value: bool,
}
```
The `deserialize_any` method looks like this, now:
```rust
    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        let peek = match self.parse_whitespace()? {
            Some(b) => b,
            None => {
                return Err(Error);
            }
        };

        if b'{' == peek {
            unimplemented!()
        } else if b'"' == peek {
            unimplemented!()
        } else if IDENTIFIER_CHARACTER_CATEGORY[peek as usize] {
            let identifier = self.parse_identifier()?;

            if !self.is_deserializing_value {
                match self.parse_whitespace()? {
                    Some(b'=') => {
                        self.next();

                        match self.parse_whitespace()? {
                            Some(_) => {
                                visitor.visit_seq(KeyValuePairDeserializer::new(self, identifier))
                            }
                            None => Err(Error),
                        }
                    }
                    _ => visitor.visit_borrowed_bytes(identifier),
                }
            } else {
                visitor.visit_borrowed_bytes(identifier)
            }
        } else {
            unimplemented!()
        }
    }
```

The way it works is as follows:
When the deserializer encounters an equals sign, it has already parsed the key and parsing it twice doesn't make sense, which is why the `KeyValuePairDeserializer` gets both the instance of the deserializer and the key. Afterwards, the `next_element_seed` method is called multiple times:
1. The state it is in is `KeyValuePairDeserializationState::DeserializedNothing`. It sets its state to `KeyValuePairDeserializationState::DeserializedKey` and calls itself with `seed.deserialize`. The implementation of `deserialize_any` simply deserializes the stored key.
2. The state it is in is `KeyValuePairDeserializationState::DeserializedKey`. It sets its state to `KeyValuePairDeserializationState::DeserializedValue` and calls the regular deserializer with `seed.deserialize`. The deserializer will then deserialize the value, while being aware, that it deserializes a value and not a key, preventing itself from accepting a faulty format like `ABC=BCD=CDE`.
3. The state it is in is `KeyValuePairDeserializationState::DeserializedValue`. The deserialization of the key-value pair is finished, so `Ok(None)` is returned to signal, that the sequence is complete.

However, as I was writing this, I realized, that both KeyValuePair and Group work with sequences for deserialization, i.e. an input like `{ABC BCD}` would theoretically also return a KeyValuePair, as it contains exactly 2 elements, which is what KeyValuePair checks for...

Therefore, I'll keep this issue open. **The "new" question is, how can I make the deserialization unambigious for this data format?** I cannot possibly work with sequences for both key-value pairs and groups.

**Comment by phlopsi** [comment]
I have solved my issue. Instead of only deserializing the key and the value, I'm also deserializing the equals sign, now.

I have copied the output of the `Deserialize` macro for `KeyValuePair` (with the help of `cargo expand`), pasted it into the module, removed the macro call and edited it to expect an equals sign in the middle. This allows me to treat sequences for key-value pairs and groups differently. Getting the rest of the deserializer to work was fairly easy and straightforward.
